File: LeetCode c#/15-3Sum.cs
//15. 3Sum

public class Solution 
{
    public IList<IList<int>> ThreeSum(int[] nums)  {
        Array.Sort(nums);

        IList<IList<int>> output = new List<IList<int>>();

        for (int i = 0; i < nums.Length - 2; i++)  {
            int left = i + 1;
            int right = nums.Length - 1;
            int sum = 0 - nums[i];

            if (i == 0 || nums[i] != nums[i - 1]) {
                while (left < right)  {
                    if (nums[left] + nums[right] == sum) {
                        output.Add(new List<int>() { nums[i], nums[left], nums[right] });
                        while (left < right && nums[left] == nums[left + 1]) { left++; }
                        while (left < right && nums[right] == nums[right - 1]) { right--; }
                        
                        left++;
                        right--;
                    }
                    else if (nums[left] + nums[right] < sum) {
                        left++;
                    }
                    else {
                        right--;
                    }
                }
            }
        }

        return output;
    }
}

File: LeetCode Java/Data Structure II/15-3Sum.java
/*
15. 3Sum
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
Notice that the solution set must not contain duplicate triplets.
Example 1:
    Input: nums = [-1,0,1,2,-1,-4]
    Output: [[-1,-1,2],[-1,0,1]]
Solution:
    So, we essentially need to find three numbers x, y, and z such that they add up to the given value.
    If we fix one of the numbers say x, we are left with the two-sum problem at hand!
*/

class Solution {
    //Time Complexity: O(n^2)
    //Space Complexity O(n) for hashset/hashmap
    public List<List<Integer>> threeSum(int[] nums) {
        
        Set<List<Integer>> res = new HashSet<>();
        Set<Integer> dups = new HashSet<>();
        Map<Integer, Integer> seen = new HashMap<>();
        
        for (int i = 0; i < nums.length; ++i) {
            
            if (dups.add(nums[i])) {
                
                for (int j = i + 1; j < nums.length; ++j) {
                    
                    int complement = -nums[i] - nums[j];
                    
                    if (seen.containsKey(complement) && seen.get(complement) == i) {
                        List<Integer> triplet = Arrays.asList(nums[i], nums[j], complement);
                        Collections.sort(triplet);
                        res.add(triplet);
                        
                    }
                    seen.put(nums[j], i);
                }
            }
        }
        return new ArrayList(res);
    }
}